File: console/shell/internal/output/output_test.go
package output

import (
	"reflect"
	"strings"
	"testing"

	"github.com/oullin/tui/internal/environment"
	"github.com/oullin/tui/internal/key"
	textutil "github.com/oullin/tui/internal/strings"
)

func withOutput(callback func(output *environment.MemoryOutput) error) error {
	output := environment.NewMemoryOutput()
	return environment.With(environment.Patch{Output: output, ErrorOutput: output}, func() error {
		return callback(output)
	})
}

func TestNoteHelpersRenderStyledLines(t *testing.T) {
	err := withOutput(func(output *environment.MemoryOutput) error {
		Note("Hello\nWorld", "")
		Warning("Careful")
		Error("Nope")
		Alert("Heads up")
		Info("Facts")
		Intro("東京\nA")
		Outro("Done")

		rendered := output.Text()
		if !strings.Contains(rendered, "Hello") || !strings.Contains(rendered, "\x1b[33m Careful\x1b[39m") || !strings.Contains(rendered, "\x1b[31m Nope\x1b[39m") {
			t.Fatalf("note output missing styled lines:\n%s", rendered)
		}
		if !strings.Contains(textutil.ParseAnsiText(rendered), "  東京 ") || !strings.Contains(textutil.ParseAnsiText(rendered), "  A    ") {
			t.Fatalf("intro output missing visible-width padding:\n%s", rendered)
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestTableGridAndTerminalHelpers(t *testing.T) {
	err := withOutput(func(output *environment.MemoryOutput) error {
		Table([]string{"Name", "Runtime"}, [][]any{{"Ollin", "OpenTUI"}})
		Grid([]any{"A", "B"}, 80)
		Title("Prompt")
		Clear()

		rendered := output.Text()
		for _, want := range []string{"Name", "Runtime", "Ollin", "OpenTUI", "┌───┬───┐", "│ A │ B │", "\x1b]0;Prompt\x07", "\x1b[H\x1b[J"} {
			if !strings.Contains(rendered, want) {
				t.Fatalf("output missing %q:\n%s", want, rendered)
			}
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestRenderGridBalancedRows(t *testing.T) {
	got := RenderGrid([]any{"component-alpha", "api-client", "theme-box", "status-task", "prompt-input", "stream-log"}, 50)
	want := strings.Join([]string{
		" ┌─────────────────┬─────────────┐",
		" │ component-alpha │ api-client  │",
		" ├─────────────────┼─────────────┤",
		" │ theme-box       │ status-task │",
		" ├─────────────────┼─────────────┤",
		" │ prompt-input    │ stream-log  │",
		" └─────────────────┴─────────────┘",
	}, "\n")
	if got != want {
		t.Fatalf("RenderGrid() =\n%s\nwant\n%s", got, want)
	}
}

func TestNotificationCommandsAndFallback(t *testing.T) {
	command := NotificationCommandFor("darwin", "Deploy", "Done", "Preview", "Glass", "")
	if command == nil || command.Bin != "osascript" || !strings.Contains(command.Args[1], `display notification "Done"`) || !strings.Contains(command.Args[1], `sound name "Glass"`) {
		t.Fatalf("darwin command = %#v", command)
	}

	commands := NotificationCommands("linux", "Deploy", "Done", "", "", "/tmp/icon.png")
	wantFirst := NotificationCommand{Bin: "notify-send", Args: []string{"--icon", "/tmp/icon.png", "Deploy", "Done"}}
	if !reflect.DeepEqual(commands[0], wantFirst) {
		t.Fatalf("linux command = %#v, want %#v", commands[0], wantFirst)
	}

	executed := []NotificationCommand{}
	ok := NotifyForPlatform("linux", "Deploy", "Done", "", "", "", NotificationRuntime{
		CommandExists: func(bin string) bool { return bin == "kdialog" },
		Execute: func(command NotificationCommand) bool {
			executed = append(executed, command)
			return true
		},
	})
	if !ok || len(executed) != 1 || executed[0].Bin != "kdialog" {
		t.Fatalf("NotifyForPlatform linux ok=%v executed=%#v", ok, executed)
	}

	err := withOutput(func(output *environment.MemoryOutput) error {
		if NotifyForPlatform("freebsd", "Deploy", "Done", "", "", "", NotificationRuntime{}) {
			t.Fatalf("unsupported notification returned true")
		}
		if !strings.Contains(output.Text(), "Deploy: Done") {
			t.Fatalf("fallback note missing:\n%s", output.Text())
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestDataTableSelectsFiltersAndCancels(t *testing.T) {
	rows := []DataTableRow{
		{Cells: map[string]any{"Name": "Alpha", "Runtime": "Node"}, Value: "alpha"},
		{Cells: map[string]any{"Name": "Beta", "Runtime": "OpenTUI"}, Value: "beta"},
	}
	err := withDataTableEnvironment(environment.NewScriptedInput([]string{key.First(key.Down), key.First(key.Enter)}), func(output *environment.MemoryOutput) error {
		result, err := DataTable(DataTableOptions{Message: "Pick project", Headers: []string{"Name", "Runtime"}, Rows: rows})
		if err != nil {
			return err
		}
		if result != "beta" {
			t.Fatalf("DataTable() = %#v", result)
		}
		if !strings.Contains(output.Text(), "Pick project") || !strings.Contains(output.Text(), "›") || !strings.Contains(output.Text(), "Bob, Developer") && !strings.Contains(output.Text(), "Beta, OpenTUI") {
			t.Fatalf("datatable output missing expected frame/summary:\n%s", output.Text())
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}

	err = withDataTableEnvironment(environment.NewScriptedInput([]string{"/", "B", key.First(key.Enter)}), func(output *environment.MemoryOutput) error {
		result, err := DataTable(DataTableOptions{Message: "Pick project", Headers: []string{"Name", "Runtime"}, Rows: rows})
		if err != nil {
			return err
		}
		if result != "beta" {
			t.Fatalf("filtered DataTable() = %#v", result)
		}
		if !strings.Contains(output.Text(), "Pick project B") {
			t.Fatalf("datatable output missing search query:\n%s", output.Text())
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}

	err = withDataTableEnvironment(environment.NewScriptedInput([]string{key.First(key.Down), key.First(key.CtrlC)}), func(output *environment.MemoryOutput) error {
		result, err := DataTable(DataTableOptions{Message: "Pick project", Headers: []string{"Name", "Runtime"}, Rows: rows})
		if err != nil {
			return err
		}
		if result != "beta" || !strings.Contains(output.Text(), "Cancelled.") || !strings.Contains(output.Text(), "/ Search") {
			t.Fatalf("cancelled DataTable() = %#v output:\n%s", result, output.Text())
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
}

func withDataTableEnvironment(input environment.Input, callback func(output *environment.MemoryOutput) error) error {
	output := environment.NewMemoryOutput()
	interactive := true
	return environment.With(environment.Patch{Input: input, Output: output, ErrorOutput: output, Interactive: &interactive}, func() error {
		return callback(output)
	})
}

File: pkgs-ts/console/tests/output.test.ts
import { describe, expect, it } from 'vitest';
import { alert, createMemoryOutput, error, grid, info, intro, note, outro, parseAnsiText, table, warning, withPromptEnvironment } from '#console/index';
import { renderGrid } from '#console/output/grid';

describe('output helpers', () => {
	it('renders note-style helpers without returning a value', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			expect(note('Hello')).toBeUndefined();
			expect(error('Nope')).toBeUndefined();
			expect(warning('Careful')).toBeUndefined();
			expect(alert('Heads up')).toBeUndefined();
			expect(info('Facts')).toBeUndefined();
			expect(intro('Start')).toBeUndefined();
			expect(outro('Done')).toBeUndefined();
		});

		expect(output.text()).toContain('Hello');
		expect(output.text()).toContain('Done');
	});

	it('renders plain multiline notes without symbol prefixes', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			note('Hello\nWorld');
		});

		expect(output.text()).toBe(' Hello\n World\n');
	});

	it('renders typed note helpers with line-oriented styling', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			warning('Careful');
			error('Nope');
			alert('Heads up');
			info('Facts');
			intro('Start\nLonger');
			outro('Done');
		});

		expect(parseAnsiText(output.text())).toBe(' Careful\n Nope\n  Heads up \n Facts\n  Start  \n  Longer \n  Done \n');
		expect(output.text()).toContain('\u001B[33m Careful\u001B[39m');
		expect(output.text()).toContain('\u001B[31m Nope\u001B[39m');
		expect(output.text()).toContain('\u001B[41m\u001B[37m Heads up \u001B[39m\u001B[49m');
		expect(output.text()).toContain('\u001B[32m Facts\u001B[39m');
		expect(output.text()).toContain('\u001B[46m\u001B[30m Start  \u001B[39m\u001B[49m');
	});

	it('pads intro and outro lines by visible width', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			intro('東京\nA');
		});

		expect(parseAnsiText(output.text())).toBe('  東京 \n  A    \n');
	});

	it('renders table and boxed grid helpers without returning a value', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			expect(table(['Name'], [['Ollin']])).toBeUndefined();
			expect(grid(['A', 'B'], 80)).toBeUndefined();
		});

		expect(output.text()).toContain('| Name');
		expect(output.text()).toContain('┌───┬───┐');
		expect(output.text()).toContain('│ A │ B │');
	});

	it('keeps grid rendering finite for invalid widths', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			expect(grid(['A', 'B'], 0)).toBeUndefined();
			expect(grid(['C', 'D'], -2)).toBeUndefined();
		});

		expect(output.text()).toContain('│ A │');
		expect(output.text()).toContain('│ B │');
		expect(output.text()).toContain('│ C │');
		expect(output.text()).toContain('│ D │');
	});

	it('normalizes fractional grid widths without overlapping boxed rows', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			expect(grid(['A', 'B', 'C', 'D'], 2.5)).toBeUndefined();
		});

		expect(output.text()).toBe(' ┌───┐\n │ A │\n ├───┤\n │ B │\n ├───┤\n │ C │\n ├───┤\n │ D │\n └───┘\n');
	});

	it('renders balanced grid rows with table separators', () => {
		expect(renderGrid(['component-alpha', 'api-client', 'theme-box', 'status-task', 'prompt-input', 'stream-log'], 50)).toBe(
			[
				' ┌─────────────────┬─────────────┐',
				' │ component-alpha │ api-client  │',
				' ├─────────────────┼─────────────┤',
				' │ theme-box       │ status-task │',
				' ├─────────────────┼─────────────┤',
				' │ prompt-input    │ stream-log  │',
				' └─────────────────┴─────────────┘',
			].join('\n'),
		);
	});

	it('does not render empty grids', async () => {
		const output = createMemoryOutput();

		await withPromptEnvironment({ output, error: output }, async () => {
			grid([]);
		});

		expect(output.text()).toBe('');
	});
});
